1단계 - 레거시 코드 리팩터링

mitsui3point opened the pull request:
안녕하세요 리뷰어님,
이번 미션은 이전 미션들보다 더 오래 걸린것 같습니다.

QnaServiceTest 의 기존 로직을 통과시키는게 조건이다 보니,
최대한 QnaServiceTest 의 기존 로직의 수정을 하지 않고 미션을 진행하려고 했는데요,
이번 미션의 목적에 맞는지 모르겠습니다..

getter 또한 완전히 없애지는 못한것 같습니다.
DeleteHistory 생성자 내부에서 getId, getWriter 들과 같은 getter 를 의존하고 있는 코드,
QnAService 에서 getId 를 의존하고 있는 findById 등..

그리고 규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
는 리팩토링하지 못했습니다.

이런 코드는 어떻게 리팩토링해 나가야하는지 이 이상 진전되지 않아 먼저 pr 드립니다.

mitsui3point committed:
docs(lms): add README_LMS

mitsui3point committed:
docs(lms): modify README_LMS

feat(lms): add Answer delete method

feat(lms): add AnswerTest delete_성공, delete_답변작성자가_아닐_경우

mitsui3point committed:
feat(lms): add Question deleteAll method

feat(lms): add AnswerTest deleteAll_성공, deleteAll_답변이_없는경우, deleteAll_다른사람이_쓴_글, deleteAll_답변_중_다른사람이_쓴_답변

mitsui3point committed:
refactor(lms): modify QnAService deleteQuestion method

refactor(lms): modify Question deleteAll method

test(lms): modify QuestionTest deleteAll_* method

refactor(lms): modify Answer delete method

test(lms): modify AnswerTest delete_* method

mitsui3point committed:
refactor(lms): modify Question unused getter/setter method

refactor(lms): modify Answer unused getter/setter method

mitsui3point committed:
refactor(lms): modify answers field type List<Answer> -> Answers (일급콜렉션 래핑)

feat(lms): add Answers add, deleteAll method

test(lms): add AnswerTest create, add_성공, deleteAll_성공, deleteAll_답변_중_다른_사람이_쓴_글

mitsui3point committed:
feat(lms): add DeleteHistory constructor
modify QuestionTest, Question, AnswerTest, Answer

test(lms): add DeleteHistoryTest create

mitsui3point committed:
feat(lms): add DeleteHistories constructor
add DeleteHistories add method
modify Answers, Answer, Question, AnswerTest, AnswersTest, QuestionTest, QnAService

test(lms): add DeleteHistoriesTest create, add_성공, value_성공, value_수정불가

dhmin5693 commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,30 +48,25 @@ public Long getId() {
+            throw new CannotDeleteException("답변작성자가 아닙니다.");
+        }
+        deleted = true;
+        deleteHistories.add(new DeleteHistory(this));
```
질문을 삭제하는데 있어 deleteHistories 가 필요하진 않습니다.
또한 매개변수인 deleteHistories에 add를 수행하게 되면 매개변수가 변경이 됩니다.
그렇게 되면 일급 함수의 장점도 잃어버리게 돼요.
일급 함수는 유지보수에 큰 장점이 있으니 한번 찾아보시길 추천드립니다.

`Answer`를 반환하기보다는 `DeleteHistories` 로 반환하는 구조로 바꿔보세요.

mitsui3point replied:
외부에서 값을 변경할수 있게 되는군요😅

이러면 내부에서 불변을 보장하는 행위가 다 의미가 없어지게 되네요.. 수정해두겠습니다

dhmin5693 commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,42 @@
+    private final List<Answer> answers;
+
+    public Answers(List<Answer> answers) {
+        this.answers = answers;
```
외부에서 이 매개변수를 변경하면 일급 컬렉션의 데이터도 변경됩니다.
[방어적 복사](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable) 문서를 확인해보세요.

mitsui3point replied:
외부에서도 수정이 가능하게 되는게 큰 문제였군요..

외부로 컬렉션을 노출시킬때 불변인것만 생각하고 있었는데, 

외부에서 주입할때 참조주소를 서로 끊어줘야 외부에서 조작할수 없게 되는건 이번에 처음 알게 되었습니다🙇

방어적 복사를 하게 되더라도 Answer 또한 외부에서 주입되고,
방어적 복사의 타겟이 List 이기 때문에,
내부 요소인 Answer들의 참조주소까지 끊기지 않아서,
내부 요소 Answer들과 외부 요소 Answer들의 참조 주소가 동일하므로,
외부 요소 Answer들에 의해 내부 요소 Answer들이 수정될 수 있으므로, 
Answer 또한 불변이어야 방어적 복사 시, List 와 Answer들 모두 불변을 보장하는것 또한 새로웠습니다😀

해결책 가이드해주셔서 감사드립니다🙇

dhmin5693 commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,42 @@
+        this.answers.add(answer);
+    }
+
+    public void deleteAll(NsUser loginUser, DeleteHistories deleteHistories) throws CannotDeleteException {
+        for (Answer answer : answers) {
+            answer.delete(loginUser, deleteHistories);
+        }
+    }
```
여기도 삭제를 하는데 `DeleteHistories`가 필요하진 않습니다.
삭제의 부산물로 `DeleteHistories`를 반환하는게 더 자연스럽지 않을까요?

mitsui3point replied:
해당 부분 수정해두었습니다🙇

dhmin5693 commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -68,21 +49,19 @@ public void addAnswer(Answer answer) {
 
-    public List<Answer> getAnswers() {
-        return answers;
+    public Question deleteAll(NsUser loginUser, DeleteHistories deleteHistories) throws CannotDeleteException {
```
여기도 다른 곳에 남긴 리뷰를 참고하여 바꿔보시기 바랍니다.

mitsui3point replied:
해당 부분 수정해두었습니다🙇

dhmin5693 reviewed (changes requested):
안녕하세요.

리팩토링 결과가 상당히 깔끔한게 눈에 잘 들어오네요.
몇 가지 아쉬운 부분에 피드백을 남겨보았으니 확인해보시고 재요청 부탁드릴게요.

이번 미션에서 필드 2개 제한은 구현하기 상당히 어려워요.
따라서 무조건 2개까지 한다고 생각하시기보다는 "최대한 줄인다" 에 중점을 두고 진행해보시면 될 것 같습니다.

어려운 내용 있으시면 언제든 DM주세요.

mitsui3point committed:
refactor(lms): modify Answer delete method(파라미터 DeleteHistories 제거, 리턴타입 Answer -> DeleteHistory 변경)

test(lms): add Answer delete_성공, delete_답변작성자가_아닐_경우

mitsui3point committed:
refactor(lms): modify Answers deleteAll,
Question delete method

test(lms): modify Answers delete_성공, delete_답변_중_다른_사람이_쓴_글
modify Question delete_성공, delete_성공_답변이_없는경우, delete_다른사람이_쓴_글, delete_답변_중_다른사람_쓴_답변

mitsui3point committed:
refactor(lms): remove DeleteHistories

mitsui3point committed:
refactor(lms): restore DeleteHistories
add DeleteHistories addAll
modify Question delete method
modify Answers delete method

test(lms): restore DeleteHistoriesTest
add DeleteHistoriesTest addAll_성공
modify QuestionTest delete tests
modify AnswersTest delete tests

mitsui3point committed:
refactor(lms): modify Answers constructor(방어적 복사 적용)

test(study): add DefensiveCopyTest
add UnmodifiableTest

dhmin5693 commented on src/main/java/nextstep/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,40 @@
+        this.deleteHistories = deleteHistories;
+    }
+
+    public DeleteHistories(DeleteHistory... deleteHistories) {
+        this(new ArrayList<>(Arrays.asList(deleteHistories)));
+    }
```
`Arrays.asList` 로 들어가서 코드를 보시면 내부에서 new ArrayList 가 이미 있는걸 볼 수 있어요.
중복으로 만들고 있네요.

그리고 자바 11에서는 List.of 생성자를 사용할 수 있습니다.
매우 편리하니 자주 사용해보세요.

mitsui3point replied:
```
    @SafeVarargs
    @SuppressWarnings("varargs")
    public static <T> List<T> asList(T... a) {
        return new ArrayList<>(a);
    }
```
중복되게 넣어두었네요 확인했습니다.!
다음반영때 수정해두겠습니다.! 감사합니다🙇

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 👍 
고생하셨습니다!

다음 단계 진행해주세요.
